### PR TITLE
Rework hammer mastery and add pummel skill

### PR DESCRIPTION
**Pull request by LordMidas**
- Rework Dented Armor to 10% less melee skill and 20% less damage dealt.
- Add new active skill Pummel.
- Hammer Mastery gives Pummel with two-handed hammers.
- Hammer Mastery causes Crush Armor and Destroy Armor to apply the Dented Armor effect.
- Rework Dent Armor perk to require blunt damage only, increase armor required to 250, and clean up code.

### CODE REVIEW

File: mod_reforged/hooks/config/strings.nut [hunk 1]
```diff
@@ -495,6 +495,24 @@ local vanillaDescriptions = [
...
+		Key = "SpecHammer",
+		Description = ::UPD.getDescription({
+			Fluff = "Master hammers and fighting against heavily armored opponents.",
+			Requirement = "Hammer",
+	 		Effects = [{
+ 				Type = ::UPD.EffectType.Passive,
+ 				Description = [
+					"Skills build up " + ::MSU.Text.colorRed("25%") + " less [Fatigue|Concept.Fatigue].",
+					"[Destroy Armor|Skill+crush_armor] and [Demolish Armor|Skill+demolish_armor_skill] inflict " + ::MSU.Text.colorGreen("33%") + " more damage against armor and apply the [Dented Armor|Skill+rf_dented_armor_effect] effect.",
+					"When using two-handed hammers gain the [Pummel|Skill+rf_pummel_skill] skill.",
```
Comment (Darxo) [inline mod_reforged/hooks/config/strings.nut:509]: This line should be under the effects category `::UPD.EffectType.Active`

File: mod_reforged/hooks/skills/perks/perk_mastery_hammer.nut [hunk 2]
```diff
@@ -0,0 +1,48 @@
+::Reforged.HooksMod.hook("scripts/skills/perks/perk_mastery_hammer", function(q) {
+	q.m.DentArmorSkills <- [
+		"actives.demolish_armor",
+		"actives.crush_armor"
+	];
+	q.m.MinArmorToDent <- 250;
```
Comment (Darxo) [inline mod_reforged/hooks/skills/perks/perk_mastery_hammer.nut:6]: How about a different name like `ArmorMaxThresholdForDent`?
The `min` in your current name is a useful information which I want to preserve by using `Threshold`. I argue that the `max` part is more important to understand what this variable does.

File: scripts/skills/actives/rf_pummel_skill.nut [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			::Time.scheduleEvent(::TimeUnit.Virtual, 1, this.onPushThrough.bindenv(this), tag);
+		}
+
+		return success;
+	}
+
+	function onPushThrough( _tag )
+	{
+		if (_tag.TargetEntity.isAlive())
+			this.line_breaker.onUse(_tag.User, _tag.TargetTile);
```
Comment (Darxo) [inline scripts/skills/actives/rf_pummel_skill.nut:114]: Linebreaker builds up 10 Fatigue on the target, this minor effect is not mentioned in the perks/skills description.

File: scripts/skills/actives/rf_pummel_skill.nut [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		}
+
+		return ret;
+	}
+
+	function isUsable()
+	{
+		return this.line_breaker.isUsable() && !this.getContainer().getActor().getCurrentProperties().IsRooted && this.getValidAttack() != null;
+	}
+
```
Comment (Darxo) [inline scripts/skills/actives/rf_pummel_skill.nut:77]: You have no `onVerifyTarget` function here.
Just like in `isUsable` you should also consult `LineBreaker` whether it's usable on a specific target.

Right now you can use Pummel on rooted or knockbackimmune enemies. LineBreaker doesn't double-check for rooted so you are theoretically able to push those enemies around with Pummel, and that doesn't seem intended (if I read the vanilla code correctly).
KnockBackImmune enemies will stop LineBreaker, but you still wasted additional fatigue in that case.

**Review by Darxo — CHANGES_REQUESTED** [review]

File: scripts/skills/actives/rf_pummel_skill.nut (hunk 4 above)
Comment (LordMidas) [reply scripts/skills/actives/rf_pummel_skill.nut:77]: This is the vanilla `onVerifyTarget` for `line_breaker.nut`. Seems like it already checks for rooted and knockback immune targets.
```squirrel
function onVerifyTarget( _originTile, _targetTile )
{
	return this.skill.onVerifyTarget(_originTile, _targetTile) && !_targetTile.getEntity().getCurrentProperties().IsRooted && !_targetTile.getEntity().getCurrentProperties().IsImmuneToKnockBackAndGrab && this.findTileToKnockBackTo(_originTile, _targetTile) != null;
}
```

File: scripts/skills/actives/rf_pummel_skill.nut (hunk 3 above)
Comment (LordMidas) [reply scripts/skills/actives/rf_pummel_skill.nut:114]: Should we remove this 10 Fatigue damage from Pummel or keep it and add it to the tooltip?

File: mod_reforged/hooks/skills/perks/perk_mastery_hammer.nut (hunk 2 above)
Comment (LordMidas) [reply mod_reforged/hooks/skills/perks/perk_mastery_hammer.nut:6]: I'm open to changing the name but `ArmorMaxThresholdForDent` is confusing in terms of wheter it is the "MaxThreshold" for "Armor" or is it the "Threshold" for "ArmorMax". Putting an underscore there could clarify it e.g. `ArmorMax_ThresholdForDent` or `Min_ArmorMax_ForDent` or `DentArmorRequiredArmorMax`. Thoughts?

Comment (Uberbagel) [reply mod_reforged/hooks/skills/perks/perk_mastery_hammer.nut:6]: Out of everything proposed, MinArmorToDent seems the best combination of lack of ambiguity and cleanliness.

File: scripts/skills/actives/rf_pummel_skill.nut (hunk 3 above)
Comment (Uberbagel) [reply scripts/skills/actives/rf_pummel_skill.nut:114]: I think we should keep it for now, it adds some nice flavor to the skill.

File: scripts/skills/actives/rf_pummel_skill.nut (hunk 4 above)
Comment (Darxo) [reply scripts/skills/actives/rf_pummel_skill.nut:77]: Ah, I overlooked that Pummel inherits from linebreaker. Forget what I said then.

File: scripts/skills/actives/rf_pummel_skill.nut [hunk 5]
```diff
@@ -34,7 +34,7 @@ this.rf_pummel_skill <- ::inherit("scripts/skills/actives/line_breaker", {
 			id = 11,
 			type = "text",
 			icon = "ui/icons/special.png",
-			text = "If the attack is successful, the target is pushed back a tile and you move into their position"
+			text = ::Reforged.Mod.Tooltips.parseString("If the attack is successful, use [Line Breaker|Skill+line_breaker] on the target")
```
Comment (Darxo) [inline scripts/skills/actives/rf_pummel_skill.nut:37]: How a bout a `for free` at the end?
`... on the target for free")`

**Review by Darxo — APPROVED** [review]
With or without my last remark, this PR is good to go

**Comment by Uberbagel** [comment]
I think the third bullet point here should come to the front for clarity.

image

**Review by Uberbagel — CHANGES_REQUESTED** [review]
### Testing and Feedback

**Dent Armor**
When an entity acts before an ally and fixes their dented armor, the entity with the dented armor still only starts with 7AP that round. I don't think this should be the case. When the Dent is fixed before the entity begins their turn.

**Pummel**
I think the third bullet point here should go to the top for clarity.

image

**Comment by Darxo** [comment]
> **Dent Armor** When an entity acts before an ally and fixes their dented armor, the entity with the dented armor still only starts with 7AP that round. I don't think this should be the case. When the Dent is fixed before the entity begins their turn.

This is happening because in Vanilla the Action Points of every entity is recovered at the same time at **round start**.

We could make it so that removing the dented effect will **always** recover Action Points on the target (even if that target is myself) equal to the MaxActionPoints that were removed by the dented effect.
That would mean that **rf_adjust_dented_armor_ally_skill** and **rf_adjust_dented_armor_skill** now recover 2 Action Points on the target/myself when used.
I kind of like that adjustment actually.

**Comment by LordMidas** [comment]
> in Vanilla the Action Points of every entity is recovered at the same time at round start.

I can see in the `actor.onTurnStart` that the Action Points of an actor are properly recovered up to the max every turn start. So I am surprised that if the Dented Armor effect is removed before the turn started, why is the character still starting their turn at 7 AP. @Uberbagel Are you sure that it wasn't the case that the dented guy had already started his turn with dented armor, then waited, then you removed the dented armor, and then when he resumed his turn you thought that he is starting?

File: scripts/skills/actives/rf_pummel_skill.nut (hunk 5 above)
Comment (LordMidas) [reply scripts/skills/actives/rf_pummel_skill.nut:37]: I'll change it to `If the attack is successful, automatically use [Line Breaker|Skill+line_breaker] for free on the target`.

**Comment by LordMidas** [comment]
In any case I just realized that we had decided in the new design to change Dented Armor to instead reduce Melee Skill by 10% and Damage by 20%. So I'll make that change and we can forget about the AP for now :P

File: scripts/skills/actives/rf_pummel_skill.nut [hunk 6]
```diff
@@ -122,8 +127,11 @@ this.rf_pummel_skill <- ::inherit("scripts/skills/actives/line_breaker", {
 		if (aoo == null)
 			return null;
 
+		if (this.m.RequiredWeaponType == null)
+			return aoo;
+
 		local weapon = aoo.getItem();
-		if (!::MSU.isNull(weapon) && weapon.isItemType(::Const.Items.ItemType.TwoHanded) && weapon.isItemType(::Const.Items.ItemType.Weapon) && weapon.isWeaponType(::Const.Items.WeaponType.Hammer))
+		if (!::MSU.isNull(weapon) && weapon.isItemType(::Const.Items.ItemType.Weapon) && weapon.isWeaponType(::Const.Items.WeaponType.Hammer))
```
Comment (Darxo) [inline scripts/skills/actives/rf_pummel_skill.nut:134]: You could use the member variable `m.RequiredWeaponType` here as it also contains `Hammer`

**Review by Darxo — CHANGES_REQUESTED** [review]

**Review by Darxo — APPROVED** [review]

**Review by Uberbagel — CHANGES_REQUESTED** [review]
Testing:
I had a freeze after killing an entity who had Dent Armor applied. This was done in Combat Simulator. The log is clean.

I'M DUMB THE BATTLE HAD ENDED BECAUSE THAT WAS THE LAST ENEMY UNIT AGGGH

**Review by Uberbagel — APPROVED** [review]
Tested. Works. Log Clean.

**Pull request merged**